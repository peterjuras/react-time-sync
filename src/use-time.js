import { useContext, useEffect, useState } from "react";
import TimeContext from "./context";

export function useTime(timerConfig = {}) {
  const timeSync = useContext(TimeContext);
  if (!timeSync) {
    throw new Error(
      "Warning! TimeSync cannot be found. Did you add <TimeProvider /> at the top of your component hierarchy?"
    );
  }

  const [time, setTime] = useState(() => timeSync.getCurrentTime(timerConfig));

  useEffect(() => timeSync.addTimer(setTime, timerConfig), [
    timerConfig.unit,
    timerConfig.interval
  ]);

  return time;
}
