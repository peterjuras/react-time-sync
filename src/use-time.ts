import { useContext, useDebugValue, useEffect, useRef, useState } from "react";
import TimeContext from "./context";
import { Interval } from "time-sync/constants";

export interface TimerConfig {
  interval?: Interval;
  unit?: number;
}

export function useTime(timerConfig: TimerConfig = {}): number {
  const timeSync = useContext(TimeContext);

  const lastConfig = useRef(timerConfig);
  const timeState = useState(() => timeSync.getCurrentTime(timerConfig));
  let [time] = timeState;
  const [, setTime] = timeState;

  if (
    timerConfig.interval !== lastConfig.current.interval ||
    timerConfig.unit !== lastConfig.current.unit
  ) {
    lastConfig.current = timerConfig;

    time = timeSync.getCurrentTime(timerConfig);
    setTime(time);
  }

  useEffect(() => timeSync.addTimer(setTime, lastConfig.current), [
    lastConfig.current.unit,
    lastConfig.current.interval
  ]);

  useDebugValue(time);
  return time;
}
