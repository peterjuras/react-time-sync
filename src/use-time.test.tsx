import { useTime, TimerConfig } from "./use-time";
import TimeSync from "time-sync";
import lolex from "lolex";
import { act, renderHook, cleanup } from "react-hooks-testing-library";

describe("#useTime", () => {
  let clock: lolex.InstalledClock<lolex.Clock>;

  beforeEach(() => {
    clock = lolex.install({ now: 1 });
  });

  afterEach(() => {
    clock.uninstall();
    cleanup();
  });

  it("should be exported correctly", () => expect(useTime).toBeDefined());

  it("should use seconds if no timeConfig is provided", () => {
    const { result, unmount } = renderHook(() => useTime());
    expect(result.current).toBe(0);
    act(() => {
      clock.tick(5000);
    });
    expect(result.current).toBe(5);
    unmount();
  });

  it("should respect prop updates", () => {
    let renderCalledCount = 0;

    const timerConfig: TimerConfig = {};
    const { result, rerender, unmount } = renderHook(() => {
      renderCalledCount++;
      return useTime({ ...timerConfig });
    });

    expect(result).toMatchSnapshot();

    act(() => {
      clock.tick(999);
    });
    expect(result).toMatchSnapshot();

    act(() => {
      clock.tick(1000);
    });
    expect(result).toMatchSnapshot();

    timerConfig.unit = 5;
    timerConfig.interval = TimeSync.MINUTES;

    rerender();

    expect(result).toMatchSnapshot();

    act(() => {
      clock.tick(1000 * 5 * 60);
    });

    expect(result).toMatchSnapshot();

    unmount();

    expect(renderCalledCount).toBe(6);
  });
});
