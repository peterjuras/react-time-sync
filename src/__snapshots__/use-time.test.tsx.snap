// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#useTime should respect prop updates 1`] = `
Object {
  "current": 0,
}
`;

exports[`#useTime should respect prop updates 2`] = `
Object {
  "current": 1,
}
`;

exports[`#useTime should respect prop updates 3`] = `
Object {
  "current": 2,
}
`;

exports[`#useTime should respect prop updates 4`] = `
Object {
  "current": 0,
}
`;

exports[`#useTime should respect prop updates 5`] = `
Object {
  "current": 300,
}
`;
